tony-core/src/main/java/com/linkedin/tony/client/TaskUpdateListener.java

```diff
@@ -0,0 +1,14 @@
+/**
+ * Copyright 2019 LinkedIn Corporation. All rights reserved. Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.tony.client;
+
+import com.linkedin.tony.rpc.TaskInfo;
+
+import java.util.Set;
+
+public interface TaskUpdateListener {
+    // Called when TonyClient gets a set of taskUrls from TonyAM.
+    public void onTaskInfosReceived(Set<TaskInfo> taskInfoSet);
```

Comment:
Can we combine `TaskUpdateListener` with `CallbackHandler` and just have one callback interface instead of two?

Reply:
They are different, CallbackHandler is something set up in the constructor to receive all one time notifications. for TaskUpdateListener, it is more like checking a stream of updates.

Reply:
I think we only call this callback once, when all the tasks have registered, so it seems a one-time notification as well.

Also, it seems that in your PR, no code is calling `onTaskInfosReceived`?

Reply:
Nevermind, I couldn't find it via Ctrl-F because GitHub had automatically collapse the `TonyClient` changes because the diff was considered "large".

Reply:
no.. every time the taskInfo is updated, we send notification

Reply:
True, the code calls `onTaskInfosReceived` every time there's a diff, but in practice, a change happens only one time:

```
// in ApplicationMaster.getTaskInfos()
      if (!singleNode && session != null && session.allTasksScheduled()) {
        return session.getTonyTasks().values().stream()
            .flatMap(tasks -> Arrays.stream(tasks).map(TonyTask::getTaskInfo))
            .collect(Collectors.toSet());
      }

      return Collections.emptySet();
```

We keep returning an empty set until ALL tasks have been scheduled, at which point we return a set containing all the tasks. There are no changes after that. So in effect, `onTaskInfosReceived` is a one-time callback.

Reply:
not really. the state of the task changes, if the task is changed from NEW -> RUNNING -> SUCCEEDED / FAILED, the will be triggered

Reply:
Are you sure? Can you add a print statement and see how many times onTaskInfosReceived() is called in a normal TonY job? Also, if it's intended that this method is called each time the task state changes, I think a better name would be `onTaskInfosChanged`.

Reply:
yah.. only when changed. That is a good point, **onTaskInfosChanged** is more proper.